src/mpi/misc/utils.c

```diff
@@ -40,43 +44,38 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
         copy_sz = sdata_sz;
 
     /* Builtin types is the common case; optimize for it */
-    if ((HANDLE_IS_BUILTIN(sendtype)) && HANDLE_IS_BUILTIN(recvtype)) {
-        MPIR_Memcpy(recvbuf, sendbuf, copy_sz);
-        goto fn_exit;
-    }
-
     MPIR_Datatype_iscontig(sendtype, &sendtype_iscontig);
     MPIR_Datatype_iscontig(recvtype, &recvtype_iscontig);
 
     MPIR_Type_get_true_extent_impl(sendtype, &sendtype_true_lb, &true_extent);
     MPIR_Type_get_true_extent_impl(recvtype, &recvtype_true_lb, &true_extent);
 
-    if (sendtype_iscontig && recvtype_iscontig) {
-#if defined(HAVE_ERROR_CHECKING)
-        MPIR_ERR_CHKMEMCPYANDJUMP(mpi_errno,
-                                  ((char *) recvbuf + recvtype_true_lb),
-                                  ((char *) sendbuf + sendtype_true_lb), copy_sz);
-#endif
-        MPIR_Memcpy(((char *) recvbuf + recvtype_true_lb),
-                    ((char *) sendbuf + sendtype_true_lb), copy_sz);
+    if (recvtype_iscontig) {
+        MPI_Aint actual_pack_bytes;
+        MPIR_Typerep_pack(sendbuf, sendcount, sendtype, 0, (char *) recvbuf + recvtype_true_lb,
+                          copy_sz, &actual_pack_bytes);
+        MPIR_ERR_CHKANDJUMP(actual_pack_bytes != copy_sz, mpi_errno, MPI_ERR_TYPE,
+                            "**dtypemismatch");
     } else if (sendtype_iscontig) {
         MPI_Aint actual_unpack_bytes;
         MPIR_Typerep_unpack((char *) sendbuf + sendtype_true_lb, copy_sz, recvbuf, recvcount,
                             recvtype, 0, &actual_unpack_bytes);
         MPIR_ERR_CHKANDJUMP(actual_unpack_bytes != copy_sz, mpi_errno, MPI_ERR_TYPE,
                             "**dtypemismatch");
-    } else if (recvtype_iscontig) {
-        MPI_Aint actual_pack_bytes;
-        MPIR_Typerep_pack(sendbuf, sendcount, sendtype, 0, (char *) recvbuf + recvtype_true_lb,
-                          copy_sz, &actual_pack_bytes);
-        MPIR_ERR_CHKANDJUMP(actual_pack_bytes != copy_sz, mpi_errno, MPI_ERR_TYPE,
-                            "**dtypemismatch");
     } else {
-        char *buf;
         intptr_t sfirst;
         intptr_t rfirst;
 
-        MPIR_CHKLMEM_MALLOC(buf, char *, COPY_BUFFER_SZ, mpi_errno, "buf", MPL_MEM_BUFFER);
+        MPL_gpu_query_pointer_type(sendbuf, &send_type);
+        MPL_gpu_query_pointer_type(recvbuf, &recv_type);
+
+        if (send_type == MPL_GPU_POINTER_DEV && recv_type == MPL_GPU_POINTER_DEV) {
+            MPL_gpu_malloc((void **) &buf, COPY_BUFFER_SZ);
+        } else if (send_type == MPL_GPU_POINTER_DEV || recv_type == MPL_GPU_POINTER_DEV) {
+            MPL_gpu_malloc_host((void **) &buf, COPY_BUFFER_SZ);
```

Comment:
This doesn't look right.  We only need the host buffer in the case where the send type is noncontiguous and is sitting on the host.

Reply:
This branch is for non-contiguous send and recv types. If either one of the two is in GPU memory we need to allocate host registered memory. Otherwise pack (unpack) will do additional copy to (from) OS bounce buffer. Am I recollecting correctly?

Reply:
Hmm.  Maybe you are right.  I miscalculated.  We probably do need both a GPU buffer and host buffer.